File: Problems/74_search2DMatrix.ts
function searchMatrix(matrix: number[][], target: number): boolean {
    //treat matrix as a sorted list
    
    const length = matrix.length * matrix[0].length;
    const width = matrix[0].length;

    let [left, right] = [0, length - 1];
    while (left <= right) {
        const mid = (left + right + 1) >> 1;
        const [referenceRow, referenceCol] = IndexConverter(mid, width);
        const reference = matrix[referenceRow][referenceCol];

        if (target == reference) return true;
        else if (target < reference) right = mid - 1;
        else left = mid + 1;
    }

    return false;
};

const IndexConverter = (index: number, width: number) => {
    //e.g. in a matrix with width 4 and height 3
    //index 0: 0, 0
    //index 5: 5 // 4, 5 % 4 -> 1, 1
    return [(index / width) >> 0, index % width];
}

console.log(searchMatrix([[1]], 1));

File: Problems/74_search2DMatrix.py
class Solution(object):
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        
        #observe that matrix is really just a sorted array split into rows
        #using regular left and right pointers, we can use regular binary search but we just need to adjust how we access the array

        matrixHeight = len(matrix)
        matrixWidth = len(matrix[0])

        left = 0
        right = matrixWidth * matrixHeight - 1 #row length * number of rows - 1 (for 0-index)

        while left < right:
            midpoint = (left + right) // 2

            if target > matrix[midpoint // matrixWidth][midpoint % matrixWidth]: #discard left 'side'
                left = midpoint + 1
            else: #discard right side but keep midpoint
                right = midpoint
        
        return True if matrix[left // matrixWidth][left % matrixWidth] == target else False
                